url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/gui/CameraControllerAquynza.java

```diff
@@ -0,0 +1,438 @@
+    {
+        //------------------------------------------------------------
+        int deltaX;
+        int deltaY;
+        boolean updated = false;
+        double senseFactor = deltaMov/5;
+
+        deltaX = e.getX() - oldMouseX;
+        deltaY = e.getY() - oldMouseY;
+
+        if ( deltaX > 5 ) deltaX = 5;
+        if ( deltaX < -5 ) deltaX = -5;
+        if ( deltaY > 5 ) deltaY = 5;
+        if ( deltaY < -5 ) deltaY = -5;
+
+        //------------------------------------------------------------
+        Matrix4x4 R; // Camera rotation matrix
+        Matrix4x4 DR;
+        Vector3D eyePosition;
+        Vector3D focusedPosition;
+        double ax, ay;
+
+        // Obtain a copy of the camera's internal parameters
+        eyePosition = camera.getPosition();
+        focusedPosition = camera.getFocusedPosition();
+
+        Vector3D u, v, w;
+        int modifiers;
+
+        modifiers = e.getModifiersEx();
+
+        R = camera.getRotation();
+        u = new Vector3D(R.M[0][0], R.M[1][0], R.M[2][0]);
+        v = new Vector3D(R.M[0][1], R.M[1][1], R.M[2][1]);
+        w = new Vector3D(R.M[0][2], R.M[1][2], R.M[2][2]);
+
+        if ( (modifiers & e.BUTTON1_DOWN_MASK) != 0 ) {
+            // Turn
+            ax = -Math.min(2, 0.01*deltaX);
+            ay = Math.min(2, 0.01*deltaY);
+
+            DR = new Matrix4x4();
+            DR.axisRotation(ay, v.x, v.y, v.z);
+            R = DR.multiply(R);
+
+            DR.axisRotation(ax, w.x, w.y, w.z);
+            R = DR.multiply(R);
+
+            updated = true;
+        }
+        else if ( (modifiers & e.BUTTON2_DOWN_MASK) != 0 ) {
+            // Move
+            eyePosition = eyePosition.substract(v.multiply(senseFactor*((double)deltaX)));
+            eyePosition = eyePosition.substract(w.multiply(senseFactor*((double)deltaY)));
+            focusedPosition = focusedPosition.substract(v.multiply(senseFactor*((double)deltaX)));
+            focusedPosition = focusedPosition.substract(w.multiply(senseFactor*((double)deltaY)));
+            updated = true;
+        }
+        else if ( (modifiers & e.BUTTON3_DOWN_MASK) != 0 ) {
+            // Advance
+            eyePosition = eyePosition.substract(u.multiply(senseFactor*((double)deltaY)));
+            ax = Math.min(2, 0.01*deltaX);
+            DR = new Matrix4x4();
+            DR.axisRotation(ax, u.x, u.y, u.z);
+            R = DR.multiply(R);
+            updated = true;
+        }
+
+        // Update camera's internal parameters from local copy
+        //R.eulerAnglesRotation(yaw, pitch, roll);
+        camera.setPosition(eyePosition);
+        camera.setFocusedPositionMaintainingOrthogonality(focusedPosition);
+        camera.setRotation(R);
+
+        //------------------------------------------------------------
+        oldMouseX = e.getX();  
+        oldMouseY = e.getY();
+        return updated;
+    }
+
+    public boolean processMouseWheelEventAwt(MouseWheelEvent e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.